samples/basic/button/src/main.c

```diff
@@ -13,89 +14,142 @@
-		printk("Error: didn't find %s device\n",
-			DT_GPIO_LABEL(DT_ALIAS(sw0), gpios));
+	button = device_get_binding(SW0_GPIO_LABEL);
+	if (button == NULL) {
+		printk("Error: didn't find %s device\n", SW0_GPIO_LABEL);
 		return;
 	}
 
-	ret = gpio_pin_configure(dev_button, DT_GPIO_PIN(DT_ALIAS(sw0), gpios),
-				 SW0_FLAGS | GPIO_INPUT);
+	ret = gpio_pin_configure(button, SW0_GPIO_PIN, SW0_GPIO_FLAGS);
 	if (ret != 0) {
-		printk("Error %d: failed to configure pin %d '%s'\n",
-			ret, DT_GPIO_PIN(DT_ALIAS(sw0), gpios),
-			DT_LABEL(DT_ALIAS(sw0)));
+		printk("Error %d: failed to configure %s pin %d\n",
+		       ret, SW0_GPIO_LABEL, SW0_GPIO_PIN);
 		return;
 	}
 
-	ret = gpio_pin_interrupt_configure(dev_button,
-					   DT_GPIO_PIN(DT_ALIAS(sw0), gpios),
+	ret = gpio_pin_interrupt_configure(button,
+					   SW0_GPIO_PIN,
 					   GPIO_INT_EDGE_TO_ACTIVE);
 	if (ret != 0) {
-		printk("Error %d: failed to configure interrupt on pin %d '%s'\n",
-			ret, DT_GPIO_PIN(DT_ALIAS(sw0), gpios),
-			DT_LABEL(DT_ALIAS(sw0)));
+		printk("Error %d: failed to configure interrupt on %s pin %d\n",
+			ret, SW0_GPIO_LABEL, SW0_GPIO_PIN);
 		return;
 	}
 
-	gpio_init_callback(&button_cb_data, button_pressed,
-			   BIT(DT_GPIO_PIN(DT_ALIAS(sw0), gpios)));
-	gpio_add_callback(dev_button, &button_cb_data);
+	gpio_init_callback(&button_cb_data, button_pressed, BIT(SW0_GPIO_PIN));
+	gpio_add_callback(button, &button_cb_data);
+	printk("Set up button at %s pin %d\n", SW0_GPIO_LABEL, SW0_GPIO_PIN);
 
-#if DT_NODE_HAS_PROP(DT_ALIAS(led0), gpios)
-	struct device *dev_led;
+	led = initialize_led();
 
-	dev_led = device_get_binding(DT_GPIO_LABEL(DT_ALIAS(led0), gpios));
-	if (dev_led == NULL) {
-		printk("Error: didn't find %s device\n",
-			DT_GPIO_LABEL(DT_ALIAS(led0), gpios));
-		return;
+	printk("Press the button\n");
+	while (1) {
+		match_led_to_button(button, led);
+		k_msleep(SLEEP_TIME_MS);
 	}
+}
 
-	ret = gpio_pin_configure(dev_led, DT_GPIO_PIN(DT_ALIAS(led0), gpios),
-				 LED0_FLAGS | GPIO_OUTPUT);
-	if (ret != 0) {
-		printk("Error %d: failed to configure pin %d '%s'\n",
-			ret, DT_GPIO_PIN(DT_ALIAS(led0), gpios),
-			DT_LABEL(DT_ALIAS(led0)));
-		return;
-	}
-#endif
-	printk("Press %s on the board\n", DT_LABEL(DT_ALIAS(sw0)));
+/*
+ * The led0 devicetree alias is optional. If present, we'll use it
+ * to turn on the LED whenever the button is pressed.
+ */
 
-	while (1) {
-#if DT_NODE_HAS_PROP(DT_ALIAS(led0), gpios)
-		bool val;
+#define LED0_NODE	DT_ALIAS(led0)
 
-		val = gpio_pin_get(dev_button, DT_GPIO_PIN(DT_ALIAS(sw0), gpios));
-		gpio_pin_set(dev_led, DT_GPIO_PIN(DT_ALIAS(led0), gpios), val);
-		k_msleep(SLEEP_TIME_MS);
+#if DT_HAS_NODE(LED0_NODE) && DT_NODE_HAS_PROP(LED0_NODE, gpios)
+#define LED0_GPIO_LABEL	DT_GPIO_LABEL(LED0_NODE, gpios)
+#define LED0_GPIO_PIN	DT_GPIO_PIN(LED0_NODE, gpios)
+#define LED0_GPIO_FLAGS	(GPIO_OUTPUT | FLAGS_OR_ZERO(LED0_NODE))
 #endif
+
+#ifdef LED0_GPIO_LABEL
+static struct device *initialize_led(void)
```

Comment:
Why not move this functions to the top and avoid forward declarations?

Reply:
Because it's optional functionality for a 'basic' sample, and I wanted to make the important parts easier for new users to find.